add uuid and usb booting with fallback to initramfs

TermeHansen opened the pull request:
As per suggestion, I have tested (on my rpi4):

-  Set up a standard installation of OSMC on an SD card.
-  Build your kernel image and use sudo dpkg -i on the image package.
-  Verify SD boot is still functional

Then for verification of USB:

-  dd the entire SD card to a USB stick or disk
-  change cmdline.txt to root=/dev/sda2 to verify USB boots OK
 

Now for UUID changes:
-  Change cmdline.txt to root=UUID=xyz to match UUID of the ext4 partition
-  Verify if changes you made to the initramfs are working as expected

Which all works as intended. The initial PR (https://github.com/osmc/osmc/pull/702#issuecomment-1701334352) also had PARTUUID included, but this does not work with busybox findfs (https://forums.gentoo.org/viewtopic-t-1095984-start-10.html)

A not that `/etc/fstab` also needs change of the mount of /boot partition to follow `cmdline.txt`

TermeHansen committed:
add uuid and usb booting with fallback to initramfs

TermeHansen committed:
add uuid and usb booting with fallback to initramfs

samnazarko commented:
I will check this shortly.
Has this been tested?

Sam

On 3 Sep 2023, at 15:21, Rasmus Lundsgaard ***@***.***> wrote:

﻿
________________________________

You can view, comment on, or merge this pull request online at:

  https://github.com/osmc/osmc/pull/745

Commit Summary

  *   cbda1b6<https://github.com/osmc/osmc/pull/745/commits/cbda1b66b1e3fac60eb8f2d6ce0218684c467044> add uuid and usb booting with fallback to initramfs

File Changes

(4 files<https://github.com/osmc/osmc/pull/745/files>)

  *   M package/kernel-osmc/build.sh<https://github.com/osmc/osmc/pull/745/files#diff-43eb2234e52017bbf63f4d3dc857ffaeb2e535f13b81bd140c86128a2b5dadce> (4)
  *   M package/kernel-osmc/initramfs-src/init<https://github.com/osmc/osmc/pull/745/files#diff-4df6270ce6bf31c33f121fdf9e072e29c073d395b3f994b37318747283fcf907> (40)
  *   M package/kernel-osmc/patches/rbp2-000-add-kernel-config.patch<https://github.com/osmc/osmc/pull/745/files#diff-b98895954d270f5e8407e47ea8335a099f514df7ee42a62dde47281c4f949526> (8)
  *   M package/kernel-osmc/patches/rbp464-000-add-kernel-config.patch<https://github.com/osmc/osmc/pull/745/files#diff-8ddf3f02c518a951495ee35b01d37865580a4626b3c10ed9a607fdf65966d76b> (6)

Patch Links:

  *   https://github.com/osmc/osmc/pull/745.patch
  *   https://github.com/osmc/osmc/pull/745.diff

—
Reply to this email directly, view it on GitHub<https://github.com/osmc/osmc/pull/745>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAHDWLBUFENEQKDHB7HCKQDXYSG6RANCNFSM6AAAAAA4JLHRJA>.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

TermeHansen commented:
I'm doing one last fresh build now of this cleaned branch, but yes I have tested all options - on rpi4. I have a rpi3, should I test on that also?

samnazarko commented:
Should be the same. Did you incorporate changes from Graham’s PR?

On 3 Sep 2023, at 15:38, Rasmus Lundsgaard ***@***.***> wrote:

﻿

I'm doing one last fresh build now of this cleaned branch, but yes I have tested all options - on rpi4. I have a rpi3, should I test on that also?

—
Reply to this email directly, view it on GitHub<https://github.com/osmc/osmc/pull/745#issuecomment-1704322807>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAHDWLDZARGDWO3JPF3UHQTXYSI6VANCNFSM6AAAAAA4JLHRJA>.
You are receiving this because you commented.Message ID: ***@***.***>

TermeHansen commented:
well added the changes to the rbp464 kernel patch, and as I wrote in the top I also removed PARTUUID (doesn't work in busybox) and added /dev/sdx to also be included in the usb wait loop - plus a bit of refactor of echo output. Nothing big...

samnazarko commented:
I suspect PARTUUID is not working because we use an MSDOS partition table and not GPT. My understanding is it is only supported in the latter. UUID should be perfectly fine however.

TermeHansen commented:
All good. Just tested the fehs build from this branch both from SD and usb.

We should also change the `/etc/fstab` entry for `/boot` to use `LABEL=OSMCInstall` as you suggested, but I think that should go in the installer fix PR.

samnazarko commented:
I've just found out that uClibC is completely broken for Aarch64, so we will need to move to GLibc. I'm seeing if that can be done without any major problems.

samnazarko commented:
We need to move to 6.1 kernel for Raspberry Pi as a priority, so that will need to come first. But I did some thinking about the next steps that should be undertaken:

1. Check (and modify) /etc/fstab and verify that the /boot partition can be mounted using UUID
2. Remove the option for USB boot in the host installer. We will then just install the root filesystem to the device that it is running off of. This should simply the configuration and avoid confusion.
3. On boot, we need to check if our OSMCInstaller partition label is based on /dev/sda or /dev/mmcblk. We then set this base partition and then create the second partition accordingly.
4. Configure fstab using UUID label of OSMCInstaller partition
5. Configure cmdline.txt using UUID label of newly created ext4 partition

Some extras that would be good:

1. We should shrink the FAT32 partition. It only needed to be of the size it was to accomodate the initial root filesystem tarball.
2. We should rename the FAT32 partition to 'OSMCBoot' so that a user does not accidentally put the USB in a PC and assume it's only the remnants of the installer. 

The installer is due a general cleanup, but this should suffice for now.

samnazarko commented:
Looks like I definitely need to move to GLibc. I will verify the targets my side so you don’t need to retest that.

I’ll then make the newer installer the default shortly.

Can I split up some of your commits like the symlink fix?

Regarding your kernel configuration changes: are you sure they are necessary. The reason being is that even when booting from SD to later mount root on USB, our initramfs doesn’t contain any kernel modules, so we should have everything we need to mount in the kernel already

Sam

On 3 Sep 2023, at 20:36, Rasmus Lundsgaard ***@***.***> wrote:

﻿

All good. Just tested the fehs build from this branch both from SD and usb.

We should also change the /etc/fstab entry for /boot to use LABEL=OSMCInstall as you suggested, but I think that should go in the installer fix PR.

—
Reply to this email directly, view it on GitHub<https://github.com/osmc/osmc/pull/745#issuecomment-1704384306>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAHDWLC3ELNJWQJJMUQHXWTXYTL2ZANCNFSM6AAAAAA4JLHRJA>.
You are receiving this because you commented.Message ID: ***@***.***>

TermeHansen commented:
Anything to help the forward movement.

I agree with your suspicion on the kernel config changes, but just added them on from the earlier PR. I will test without on my rpi4 and get back here.

TermeHansen commented:
Anything to help the forward movement.

I agree with your suspicion on the kernel config changes, but just added them on from the earlier PR. I will test without on my rpi4 and get back here.

On another node, the "qemu" dummy package is left out on Debian bookworm, so had to change that dependency in the toolchain. But I left that out from here - just FYI...

graham8 commented:
It's good this is getting some love now as I never found time to follow up with changes to the installer. I tested my patches a lot on RPi3B but I don't have a Pi4.

@samnazarko I'm confused as to why you would remove the USB boot option from the installer.  Isn't that exactly what people want?  Shouldn't they just be able to write to a USB and expect that to work? Or is the idea the storage format becomes irrelevant?

Sooner or later, you will find a USB device that you can't boot from because the driver is not in the bootloader/initramfs.  I did include one driver which IIRC let me boot from a SD card in a USB SD card reader. But I couldn't boot from a spinning rust drive in a cheap USB-IDE caddy and failed to find the necessary modules. A friend boots Manjaro on RPi4 from a SSD in a USB-SATA caddy but I haven't been able to test that with OSMC.

TermeHansen commented:
With this we don't need to define if it is a SD or USB drive you are installing from...

I'm booting from a SSD via usb on my rpi4 now with no SD needed.

samnazarko commented:
Well that's weird. I was chasing down an issue on Vero (ucLibc) and noticed that installation wasn't working because mkfs.ext4 has moved from /usr/sbin/ to /sbin. I'm not sure why that didn't cause a problem on RBP..

TermeHansen committed:
remove rbp4 patches as not neeed

TermeHansen commented:
Just tried a build without the kernel config patches and it still worked fine on my rpi4. I could try the same with a rbp2 build on my rpi3 if you think @samnazarko  Also I might also test with a usb hub and another en drive also...

TermeHansen commented:
@samnazarko what is the plan with change? I would really like to get this in, as it is already living on my production machine now 🤗

samnazarko commented:
Hopefully after today's announcement it is clear why my focus is to get 6.1 out of the door now :) 

Then I'll get back to this.

samnazarko commented:
First thing I've done is adjust the symlink in the kernel build script as per your PR.

I will look at the initramfs changes shortly.
Any kernel changes needed can come with 6.1.

TermeHansen commented:
@samnazarko What is the outline of this getting into master? I rebased my PR branch with master last night, but somehow github likes to show all the commits still, so maybe better to make a new PR when the time is ready to test on master and get in?

TermeHansen committed:
remove rbp4 patches as not neeed (again)

samnazarko commented:
Might be best to make new PR but it’s good to keep comments from original so we remember. Maybe open a new PR when we are ready for merge

On 12 Dec 2023, at 17:09, Rasmus Lundsgaard ***@***.***> wrote:

﻿

@samnazarko<https://github.com/samnazarko> What is the outline of this getting into master? I rebased my PR branch with master last night, but somehow github likes to show all the commits still, so maybe better to make a new PR when the time is ready to test on master and get in?

—
Reply to this email directly, view it on GitHub<https://github.com/osmc/osmc/pull/745#issuecomment-1852457648>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAHDWLA3JNQJ7XJ2LRFDNU3YJCFVTAVCNFSM6AAAAAA4JLHRJCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNJSGQ2TONRUHA>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

TermeHansen commented:
> Might be best to make new PR but it’s good to keep comments from original so we remember. Maybe open a new PR when we are ready for merge 

I solved it by `git pull --rebase upstream master` instead of to my forked master :smile:

samnazarko commented:
That's good!

SpectralCascade commented:
Is this feature expected to be integrated as part of the Linux 6.1 kernel / pi 5 release?

samnazarko commented:
6.6 and when it’s ready

samnazarko commented:
I agree that we should get this merged sooner than later
I'll look over the weekend and come up with some suggestions.

TermeHansen committed:
Merge branch 'osmc:master' into rbp_usb_boot

TermeHansen commented:
kind reminder @samnazarko - now building it again locally :smile:

TermeHansen committed:
Merge branch 'osmc:master' into rbp_usb_boot
